fix: Allow companion functions when result type is not resolvable given intermediate type

rui-mo opened the pull request:
The registrations of partial and merge companion functions does not require the 
result type is resolvable given intermediate type. This PR removes the 
limitations for them.The registration of merge_extract companion function used 
to depend on the resolving, and this PR uses the passed-in result type rather 
than the resolved from intermediate type.

zhztheplayer commented:
@rui-mo Thank you for taking this on. I remember we need to bypass [this check](https://github.com/facebookincubator/velox/blob/9a4519f9a87f91ff641a8b812811b76d800bb2fb/velox/exec/AggregateCompanionAdapter.cpp#L394-L397) as well to make the decimal aggregate functions work. Or we don't need that change anymore?

rui-mo commented:
@zhztheplayer Thanks for the pointer! In Gluten we expect that all the names will end with "_merge_extract," however according to this reasoning, the names may have other suffixes. For instance, we would obtain the names for sum and average as follows. I'll investigate further to see how to incorporate this section with Gluten.

> avg_merge_extract_DECIMAL, avg_merge_extract_double
> sum_merge_extract_bigint, sum_merge_extract_DECIMAL, sum_merge_extract_double

zhztheplayer commented:
> @zhztheplayer Thanks for the pointer! In Gluten we expect that all the names will end with "_merge_extract," however according to this reasoning, the names may have other suffixes. For instance, we would obtain the names for sum and average as follows. I'll investigate further to see how to incorporate this section with Gluten.
> 
> > avg_merge_extract_DECIMAL, avg_merge_extract_double
> > sum_merge_extract_bigint, sum_merge_extract_DECIMAL, sum_merge_extract_double

Sounds reasonable. Look forward to a solution here. Thanks!

zhztheplayer commented on velox/exec/AggregateCompanionAdapter.cpp:
```diff
@@ -340,20 +340,11 @@ bool registerAggregateFunction(
                      core::AggregationNode::Step::kFinal,
                      argTypes,
-                     originalResultType,
+                     resultType,
```
> Suppose an aggregation node perform the intermediate aggregation step of the merge-extract-function, then both the argTypes and the resultType received by the factory at line 337 are the intermediate type of the original function.

Can we break down the cases here? E.g., when step is `single` / `final`, we just skip the result type resolution?

rui-mo replied:
> Can we break down the cases here? E.g., when step is single / final, we just skip the result type resolution?

@kagamiori I apologize for the delayed response; I was on vacation. Do you believe that @zhztheplayer's suggestion above makes sense? We are proposing this change to allow more flexibility in the aggregate function registration especially for the Spark decimal average.

NEUpanning replied:
I am curious why the original function factory is called with`Step::kFinal`. If the step is eithter partial or intermediate, could we use `Step::kIntermediate` instead? In this way, there is no need for the result type resolution.

NEUpanning replied:
I am curious why the original function factory is called with`Step::kFinal`. If the step is eithter partial or intermediate, could we use `Step::kIntermediate` instead? In this way, there is no need for the result type resolution.

rui-mo replied:
@NEUpanning I think using `kIntermediate` cannot ensure the result could be correctly extracted if the `MergeExtractFunction::extractValues` is called.

NEUpanning replied:
@rui-mo I am not sure I follow. Could you elaborate on it? I assume that for partial and intermediate steps, the input and result types of merge_extract are both intermediate type, which maps to **intermediate step** of original aggregate function. Thanks!

rui-mo replied:
@NEUpanning With the change as you mentioned, it appears we are only registering a function accepting intermediate input and giving intermediate output for the partial/intermediate steps (this is the same with an ordinary aggregate function). But for a companion function, we need to register a function accepting intermediate input and can give final output for the partial/intermediate steps. @kagamiori Would you like to confirm? Thanks.

NEUpanning replied:
> for a companion function, we need to register a function accepting intermediate input and can give final output for the partial/intermediate steps

@rui-mo Thanks for your clarification! For the partial/intermediate steps, I thought aggregate op would call `extractAccumulators` function to get results as stated in [the document ](https://facebookincubator.github.io/velox/develop/aggregate-functions.html#groupby-aggregation), and `merge_extract` does not override `extractAccumulators` to `extractValues`. Could you remind me how aggregate op gets the final output for the partial/intermediate steps?

rui-mo replied:
I got your point. By reviewing the design doc https://github.com/facebookincubator/velox/discussions/4493 I'm not sure if merge_extract is designed to extract final result when step is partial//intermediate. Perhaps @kagamiori could help clarify. Thanks!

NEUpanning commented:
@rui-mo Thank you for working on this. I'm developing a new Spark `collect_list` with the signature `T -> varbinary -> array(T)`, for more details see #12023. I suppose if this PR is merged, Velox would support this signature. So, I cherry-picked this PR into my branch and ran the test. I found three issues:
1. [`CompanionSignatures::extractFunctionSignature`](https://github.com/facebookincubator/velox/blob/main/velox/exec/AggregateCompanionSignatures.cpp#L212) still has limitations. It leads to register no extract functions.
2. While creating the merge_extract signature, Velox checks if all type variables should appear in the inputs arguments. For this signature, `T` does not appear in `varbinary`.  See [code link](https://github.com/facebookincubator/velox/blob/main/velox/expression/FunctionSignature.cpp#L161)
3. `exec::test::PlanBuilder::project` uses [resolveScalarFunctionType](https://github.com/facebookincubator/velox/blob/main/velox/parse/TypeResolver.cpp#L80) to resolve result type from intermediate type, but it cannot be resolved for this signature. Here is the error message:
```
C++ exception with description "Exception: VeloxUserError
Error Source: USER
Error Code: INVALID_ARGUMENT
Reason: Scalar function signature is not supported: spark_collect_list_extract(VARBINARY). Supported signatures: (varbinary) -> array(E).
Retriable: False
Function: resolveScalarFunctionType
File: /Users/panning/VSProjects/glutenNvelox_CE/velox2/velox/velox/parse/TypeResolver.cpp
Line: 102
```

cc @zhztheplayer

rui-mo commented:
>  I'm developing a new Spark collect_list with the signature T -> varbinary -> array(T)

Hi @NEUpanning, could you elaborate more about the issue you met when developing the collect_list, and how the issue might be relevant to this PR perhaps in the issue https://github.com/facebookincubator/velox/issues/12023? Thanks.

NEUpanning commented:
@rui-mo To resolve issue #12023, we have to change the intermediate type of collect_list to varbinary. After this change the signature of collect_list will be `T -> varbinary -> array(T)` that leads to result type is not resolvable given intermediate type and generates no companion functions for collect_list. I thought this PR would allow companion functions for the new collect_list signature.

kagamiori commented:
Hi @rui-mo and @zhztheplayer, sorry for the delay.
> Can we break down the cases here? E.g., when step is single / final, we just skip the result type resolution?

This could be one possibility, but I'd like to understand the problem a bit better. Below are the problems I got from the conversation here:
1. @rui-mo needs a workaround to allow registering _merge_extract functions of decimal aggregation functions whose result types cannot be resolved from intermediate types.
2. Similarly, @NEUpanning needs a way to create function signature and register _extract functions of aggregation functions whose result type cannot be resolved from intermediate types, such as `T -> varbinary -> array(T)`.

These two problems are similar. Could you please help me understand how the _merge_extract and _extract function will be used in Gluten if the function signature contain unresolvable result type (in other words, where does Gluten get the correct result type)? For example, are these _merge_extract and _extract functions used directly in the user's queries? If so, how does Gluten determine the correct result type when constructing an aggregation or project plan node? Or does Gluten only allow user's queries to use the original aggregation function and the plan nodes of _merge_extract and _extract functions are created according to the plan node of the original aggregation function?

rui-mo commented:
@kagamiori Gluten uses **Velox's final aggregation + companion functions (partial, merge, merge_extract)** to map to different modes of Spark aggregation. Velox's single aggregation is also utilised for Spark's complete aggregation. The reasons to use companion functions can be referred at https://github.com/facebookincubator/velox/discussions/9048#discussioncomment-10643416. This replacement is plan-level, so the user cannot use them in queries directly, instead Gluten will generate them according to the Spark plan.

The extract companion function is not used by Gluten, and for the merge_extract function, Gluten uses the result type from Spark planner. Therefore, Gluten only needs to enable the **merge_extract function of final aggregate** when result type is not resolvable given intermediate type. Do you think it makes sense? Thanks!

rui-mo commented:
@kagamiori I updated this PR to skip type resolving only for single and final steps, and added tests for the merge_extract companion function in `testAggregationsWithCompanion`. Would you help take another look? Thanks!

kagamiori commented:
> @kagamiori Gluten uses **Velox's final aggregation + companion functions (partial, merge, merge_extract)** to map to different modes of Spark aggregation. Velox's single aggregation is also utilised for Spark's complete aggregation. The reasons to use companion functions can be referred at [#9048 (reply in thread)](https://github.com/facebookincubator/velox/discussions/9048#discussioncomment-10643416). This replacement is plan-level, so the user cannot use them in queries directly, instead Gluten will generate them according to the Spark plan.
> 
> The extract companion function is not used by Gluten, and for the merge_extract function, Gluten uses the result type from Spark planner. Therefore, Gluten only needs to enable the **merge_extract function of final aggregate** when result type is not resolvable given intermediate type. Do you think it makes sense? Thanks!

Hi @rui-mo, thank you for sharing the context! Since it is specific to Gluten that it only uses companion functions with the `single` and `final` aggregation steps and not using companion function signatures during query resolution, I think the condition should b relaxed only in Gluten. One possibility comes to my mind is for Gluten to implement its own CompanionFunctionsRegistrar that inherits the Velox one and override the registerMergeExtractFunction() method.

cc @mbasmanova for additional thoughts.

rui-mo commented:
> One possibility comes to my mind is for Gluten to implement its own CompanionFunctionsRegistrar that inherits the Velox one and override the registerMergeExtractFunction() method.

@kagamiori Thanks for your suggestion! I assume we could allow a custom `CompanionFunctionsRegistrar` to be registered by providing a factory method in Velox, and Gluten can register its customized one. Similar customization may need for the `CompanionSignatures` which skips the partial/merge/merge_extract signatures when `isResultTypeResolvableGivenIntermediateType` returns false. By doing so, Spark aggregate functions whose return type cannot be resolved given intermediate type (like decimal_avg) can only be added in Velox without the registration of companion functions. 

On the other hand, the change proposed in this PR only relax the restrictions for partial, merge, and merge_extract of single and final steps. The registrations of them does not depend on the resolvability in the first place, so I suppose this change is more like to remove the unnecessary limitation.

Could you please help provide more insights? Thanks.

kagamiori commented:
Hi @rui-mo, sorry for the delay. I looked at the code in this PR. One of my concern is about the removal of the checks `isResultTypeResolvableGivenIntermediateType(signature)` in AggregateCompanionSignatures.cpp. Previously if the UDAF function signature doesn't pass this check, it fails at registration time of its companion functions. If we remove this check, the registration would succeed but it will fail at query execution time (if merge- or merge-extract companion functions are used with the intermediate aggregation step). This seems to not be what we desire in general, so I would suggest try relaxing the check only in Gluten if that's possible.

rui-mo commented:
@kagamiori Thanks for helping review!

>  If we remove this check, the registration would succeed but it will fail at query execution time (if merge- or merge-extract companion functions are used with the intermediate aggregation step).

It looks to me the partial and merge companions do not depend on the resolving from intermediate type to result type. Please kindly correct me if that's wrong. For the merge_extract, we are not sure if it is designed to extract final result when step is partial or intermediate because it does not override the `extractAccumulators` and intermediate outputs will be produced, seeing details in https://github.com/facebookincubator/velox/pull/11999#discussion_r1985925064. If that's the case, its result type resolving could also be unnecessary. How do you think?

rui-mo commented:
Hi @mbasmanova @kagamiori, this is a summary on the scope of this PR. Could you spare some time to help review? Thanks!

---
Gluten uses Velox's final aggregation + companion functions (partial, merge, merge_extract) to map to different modes of Spark aggregation. While the required resolving from intermediate type to result type during companion functions' registration causes failure for Spark decimal_avg and collect_list aggregate functions, whose result type cannot be resolved from the intermediate type. This PR aims to relax the limitation when unnecessary.

**PR Summary**

1) Based on the implementation, the 'partial' and 'merge' companion functions do not rely on the intermediate-type-to-result-type-resolving. We remove the requirement for them.
2) For the 'merge_extract' companion function, when called with single or final step, the user-passed output type is the same with the original result type, so we could remove the resolving process for single and final steps.

With the above fixes, Gluten issue could be resolved without bringing side-effect to Velox functionality as we only remove the unnecessary resolving process.

**Further thoughts**

@NEUpanning finds the `MergeExtractFunction` does not override the `extractAccumulators` function, which means when called with 'partial' or 'intermediate' step, intermediate result is produced. This brings the question on whether the merge_extract function is designed to produce final result even for 'partial' and 'intermediate' steps. If not, `Intermediate` rather than `kFinal` shall be used during `func->factory` for 'partial' and 'intermediate' steps. If so, the intermediate-type-to-result-type-resolving could be removed completely for the merge_extract function. 

https://github.com/facebookincubator/velox/blob/8d5fa8f88e623eaf27a6186061ec4dbc8639f8fb/velox/exec/AggregateCompanionAdapter.cpp#L358-L366

Please note that although this modification is proposed based on Gluten's reliance on companion functions, if the community agrees with the enhancement proposed by @zhztheplayer in https://github.com/facebookincubator/velox/issues/12830, it could also be considered as a simplification as unnecessary limitation during the companion function registration is removed.

rui-mo commented:
@pedroerp @FelixYBW This is an essential one for Gluten. This patch and https://github.com/facebookincubator/velox/pull/12861 propose two different approaches to address the same issue we encountered while integrating Velox aggregate with Gluten. We’d appreciate Meta’s feedback on which direction Velox would prefer to take, and we’ll be happy to follow up with any further review suggestions. cc: @mbasmanova

mbasmanova commented:
@rui-mo Rui, would it be possible to extend this PR to add a test that illustrate a query plan that was not possible before the change and that is possible now?

rui-mo commented:
> @rui-mo Rui, would it be possible to extend this PR to add a test that illustrate a query plan that was not possible before the change and that is possible now?

I added a test `TEST_F(AggregationTest, companionWithUnresolvableType)`. Thanks for the suggestion.

rui-mo commented on velox/exec/AggregateInfo.cpp:
```diff
@@ -83,9 +83,12 @@ std::vector<AggregateInfo> toAggregateInfo(
+      info.intermediateType =
+          resolveAggregateFunction(
+              aggregate.call->name(), aggregate.rawInputTypes)
+              .second;
```
https://github.com/facebookincubator/velox/commit/db8a1df0cd9cae0f1197019711d02c13472383e5 assumes that the result type could always be resolved, but there are some special functions we know their result type could not be resolved given the intermediate type. I introduced the `resultTypeResolvable` option to let the planner explicitly specify it.

E.g. in the `decimal_avg` function, the resolving fails due to below exception (`a_precision` is in the raw input type, and the intermediate and final type are unaware of it).


> E20250903 17:53:44.099695 7465310 Exceptions.h:53] Line: /Users/rui/Downloads/velox-dev/./velox/expression/type_calculation/Scanner.h:59, Function:getValue, Expression: values_.count(varName) Variable a_precision is not defined, Source: RUNTIME, ErrorCode: INVALID_STATE

@mbasmanova Do you think this makes sense? Thanks.

rui-mo commented:
@mbasmanova I'd appreciate it if you could take another look, thanks!

mbasmanova commented:
@rui-mo I've been trying to understand this change... but I find it very hard. I wish there was a clear easy to follow explanation of what doesn't work today and how this change makes it work. I believe there is a particular pattern of an aggregation query that doesn't work. Can we add that as a simple test so I can run it and see how that works / doesn't work?

rui-mo commented:
@mbasmanova Sure. I'll prepare a clearer explanation and a reproducible test.

rui-mo commented:
Hi @mbasmanova, I’ve appended an explanation in response to your questions, along with a test case to help reproduce the issue. I hope this provides better clarity. Please don’t hesitate to let me know if you have any further questions or need additional details.

## What does not work

The registrations of companion functions (`partial`, `merge`, `merge_extract`) are impossible when the result type of an aggregate function cannot be resolved from the intermediate type.

For example, the signature of Spark `collect_list` function is `T -> varbinary -> array(T)`,  
where the `array(T)` cannot be resolved with `varbinary`. The Spark `decimal_avg` is `DECIMAL(a_precision, a_scale) -> ROW(DECIMAL(i_precision, a_scale), bigint) -> DECIMAL(r_precision, r_scale)`, where `r_precision` cannot be resolved from `i_precision`.

While the above functions are valid aggregate functions, they cannot be registered due to limitations in Velox. Specifically, two limitations are responsible for the registration failures.
1. Velox returns nullptr for the companions function signatures if the result type is unresolvable from intermediate type.
   Related code are `CompanionSignatures::partialFunctionSignatures`, `CompanionSignatures::mergeFunctionSignature`, `CompanionSignatures::mergeExtractFunctionSignature`.
2. Aggregate assumes that the resolve of `raw input types -> intermediate type -> result type` should always succeed. If unresolvable, an exception is thrown.
   Realted code is below.
```
info.intermediateType = resolveAggregateFunction(
                             aggregate.call->name(), aggregate.rawInputTypes)
                             .second;
```

## How this change makes it work

To address these two limitations, this change introduces the following updates to enable the registration of these functions.

1. For the first limitation, it turns out that
   - The `partial` and `merge` companion functions overrdie `extractValues` as `extractAccumulators`, so intermediate-to-result type resolving (hereafter referred as "type resolving") actually does not occur in this process.
   - For the `merge_extract` companion function, it receives intermediate inputs and returns the final result. However, when invoked in the single or final step, the user-provided output type is already the final result type. Therefore, type resolving is unnecessary and can be skipped for these two steps.
   Based on the above, we can return a valid signature for them even when the result type cannot be resolved from the intermediate type.
2. The second limitation is that Velox assumes the result type can always be resolved from the intermediate type. However, as we've already observed, there are exceptions to this assumption. To address this, I introduced the `resultTypeResolvable` option in the Aggregate function call, allowing the planner to explicitly specify whether type resolution should occur. When the planner sets this option to false, type resolving is skipped.

## Use the below test to reproduce

```
#include "velox/exec/AggregateCompanionSignatures.h"

TEST_F(AggregationTest, companionWithUnresolvableType) {
  const std::string kFunctionName = "test_aggregate";
  auto inputType = DECIMAL(18, 6);
  auto data = makeRowVector({
      makeFlatVector<int64_t>({1, 2, 3}, inputType),
  });

  registerAggregateFunction(
      kFunctionName,
      {AggregateFunctionSignatureBuilder()
           .integerVariable("a_precision")
           .integerVariable("a_scale")
           .integerVariable("i_precision", "min(38, a_precision + 10)")
           .integerVariable("r_precision", "min(38, a_precision + 4)")
           .integerVariable("r_scale", "min(38, a_scale + 4)")
           .argumentType("DECIMAL(a_precision, a_scale)")
           .intermediateType("ROW(DECIMAL(i_precision, a_scale), bigint)")
           .returnType("DECIMAL(r_precision, r_scale)")
           .build()},
      [&](core::AggregationNode::Step /*step*/,
          const std::vector<TypePtr>& /*argTypes*/,
          const TypePtr& /*resultType*/,
          const core::QueryConfig& /*config*/)
          -> std::unique_ptr<exec::Aggregate> { VELOX_NYI("Unimplemented"); },
      /*registerCompanionFunctions=*/true,
      /*overwrite=*/true);

  auto makePlan = [&](const core::CallTypedExprPtr& aggExpr) {
    return PlanBuilder()
        .values({data})
        .addNode([&](auto nodeId, auto source) -> core::PlanNodePtr {
          std::vector<TypePtr> rawInputTypes;
          for (const auto& input : aggExpr->inputs()) {
            rawInputTypes.push_back(input->type());
          }

          std::vector<core::AggregationNode::Aggregate> aggregates{
              {aggExpr,
               rawInputTypes,
               nullptr,
               {},
               {},
               /*distinct=*/false}};

          return std::make_shared<core::AggregationNode>(
              nodeId,
              core::AggregationNode::Step::kFinal,
              std::vector<core::FieldAccessTypedExprPtr>{},
              std::vector<core::FieldAccessTypedExprPtr>{},
              std::vector<std::string>{"agg"},
              aggregates,
              false,
              std::move(source));
        })
        .planNode();
  };

  CursorParameters params;
  auto intermediateType = ROW({"c0", "c1"}, {DECIMAL(28, 6), BIGINT()});
  std::vector<core::TypedExprPtr> inputs = {
      std::make_shared<core::FieldAccessTypedExpr>(DECIMAL(18, 6), "c0")};

  // Partial companion function.
  params.planNode = makePlan(
      std::make_shared<core::CallTypedExpr>(
          intermediateType,
          inputs,
          CompanionSignatures::partialFunctionName(kFunctionName)));
  readCursor(params);

  // Merge companion function.
  inputs = {
      std::make_shared<core::FieldAccessTypedExpr>(intermediateType, "c0")};
  params.planNode = makePlan(
      std::make_shared<core::CallTypedExpr>(
          intermediateType,
          inputs,
          CompanionSignatures::mergeFunctionName(kFunctionName)));
  readCursor(params);

  // Merge extract companion function.
  auto resultType = DECIMAL(22, 10);
  inputs = {
      std::make_shared<core::FieldAccessTypedExpr>(intermediateType, "c0")};
  params.planNode = makePlan(
      std::make_shared<core::CallTypedExpr>(
          resultType,
          inputs,
          CompanionSignatures::mergeExtractFunctionName(kFunctionName)));
  readCursor(params);
}
```

mbasmanova commented:
@rui-mo Rui, thank you for summarizing the issue. I have some follow-up questions.

> The registrations of aggregate companion functions (partial, merge, merge_extract) are impossible when the result type of an aggregate function cannot be resolved from the intermediate type.

Companion functions were designed to be used as standalone functions. Specifically, the use case for these is to allow one query to compute intermediate results and write them to a table, then another query to read intermediate results from multiple partitions of the table and compute final result. 

A function signature "varbinary -> array(T)" is not valid as it is not possible to use such function in a SQL query. Hence, Velox doesn't generate such companion functions. Do that make sense?

I believe you'd like to use companion functions in a different scenario where some of the above is not application. Would you describe that scenario? If possible, please, up the writeup above to keep all the relevant information in a single place. Thanks.

mbasmanova commented:
@rui-mo @zhztheplayer I apologize it took so long to make progress on this. Thank you so much for explaining the problem in detail. https://github.com/facebookincubator/velox/pull/11999#issuecomment-3274577979 combined with #12830 help me understand the context. Let me summarize my understanding to make sure I didn't miss anything.

Spark query planner relies on execution to support aggregations with a mix of partial, intermediate and final aggregate functions. #12830 shows examples how these are used and explains why these make sense.

Velox doesn't support such mixed aggregations.

Gluten came up with a workaround of using "single" Velox aggregations with companion functions. This works... except when companion functions cannot be generated because they cannot be used as standalone functions in a query.

The proposal is to change Velox to allow generating companion functions even if they cannot be used standalone with the understanding that these functions will be used as part of "single" aggregations and return types will be provided explicitly.

This makes sense to me and I'd be happy to accept the PR.

Now that the major questions are resolved, all I have is a few minor comments and requests. Please, consider these and let me know if they make sense. None of these are deal breakers for me.

(1) Would you help me understand why is it Ok to leave AggregationInfo.intermediateType field unset. Is it not being used? Is it not being used for single agg specifically? If so, can we avoid introducing a flag in the plan node and instead change resolveAggregateFunction to tryResolveAggregateFunction, which would return nullptr on failure instead of throwing?

(2) Would it be possible to add e2e tests that used aggregations from #12830 to help understand the e2e use case better?

(3) Would you add a comment next to call to tryResolveAggregateFunction and in register-companion-functions APIs to explain why we expect resolution to not succeed sometimes and why it is Ok. It would be great to reference https://github.com/facebookincubator/velox/pull/11999#issuecomment-3274577979 and #12830 

Appreciate your patience and willingness to explain. Again, sorry for such delay.

kgpai commented:
@rui-mo Apologies, can you rebase and fix conflicts. I will start the import process then.

rui-mo commented on velox/exec/Aggregate.cpp:
```diff
@@ -296,22 +296,33 @@ std::unique_ptr<Aggregate> Aggregate::create(
+      // aggregate operator rather than as a standalone function at the SQL
+      // level. In such cases, the caller should explicitly specify the result
+      // type.
+    }
```
Reverted according to https://github.com/facebookincubator/velox/commit/2e3fc87121e26aaab6bade362f6e0da47281294d.

kKPulla commented:
@rui-mo  @kgpai I see around 70+ failure signals internally blocking the merge. I will try and fix them later this week and keep you posted here on the progress.
